[Clang] suppress deprecated field warnings in implicit special-member functions

shashforge opened the pull request:
What I did

In clang/lib/Sema/SemaAvailability.cpp, under case AR_Deprecated:, I added an early-exit guard to suppress deprecation diagnostics in purely implicit, defaulted special-member functions:

```
if (const auto *FD = dyn_cast_or_null<FunctionDecl>(S.getCurFunctionDecl());
    FD && FD->isImplicit())
  return;

```

**Added clang/test/Sema/implicit-deprecated-special-member.cpp to verify:**



**Updated clang/docs/ReleaseNotes.rst under Bug Fixes to C++ Support:**
`
Suppress -Wdeprecated-declarations in implicitly-generated special-member functions for deprecated data-members; only explicit member accesses still warn. (Issue #147293)`

**Testing**
* Rebuilt Clang with `ninja clang`.
* Ran `ninja check-clang` all tests pass, including the new lit test.

---

Fixes https://github.com/llvm/llvm-project/issues/147293

Please let me know if you spot anything I missed or if there’s any further tweak needed. Thanks for your time!

shashforge committed:
Sema: suppress deprecated field warnings in implicit special-member functions

Only explicit uses of deprecated fields should warn.
Defaulted constructors and copy/move assignment operators (when
implicitly generated) no longer trigger -Wdeprecated-declarations.

* In SemaExpr.cpp, early-return in DiagnoseUseOfDecl for implicit SM.
* Add lit regression test in clang/test/SemaCXX/deprecated-no-implicit-warning.cpp.

Fixes: https://github.com/llvm/llvm-project/issues/147293

Signed-off-by: Shashi Shankar <shashishankar1687@gmail.com>

shashforge committed:
Docs: add release note for implicit-special-member deprecation fix

zwuis commented on clang/lib/Sema/SemaAvailability.cpp:
```diff
@@ -452,6 +452,28 @@ static void DoEmitAvailabilityWarning(Sema &S, AvailabilityResult K,
       }
     }
   }
+  // Suppress -Wdeprecated-declarations in purely implicit special-member functions.
+  if (K == AR_Deprecated) {
+    if (const FunctionDecl *FD = S.getCurFunctionDecl()) {
+      // Implicit, defaulted constructor (default / copy / move)
+      if (auto *CD = dyn_cast<CXXConstructorDecl>(FD)) {
+        if (CD->isImplicit() && CD->isDefaulted())
+          return;
+      }
+      //Implicit, defaulted destructor
+      if (auto *DD = dyn_cast<CXXDestructorDecl>(FD)) {
+        if (DD->isImplicit() && DD->isDefaulted())
+          return;
+      }
+      //Implicit copy- or move-assignment operator
+      if (auto *MD = dyn_cast<CXXMethodDecl>(FD)) {
+        if (MD->isImplicit() &&
+            (MD->isCopyAssignmentOperator() || MD->isMoveAssignmentOperator()))
+          return;
+      }
+    }
+  }
```
Please move into the `switch` statement below.

zwuis commented on clang/lib/Sema/SemaAvailability.cpp:
```diff
@@ -452,6 +452,28 @@ static void DoEmitAvailabilityWarning(Sema &S, AvailabilityResult K,
   }
+  // Suppress -Wdeprecated-declarations in purely implicit special-member functions.
+  if (K == AR_Deprecated) {
+    if (const FunctionDecl *FD = S.getCurFunctionDecl()) {
+      // Implicit, defaulted constructor (default / copy / move)
+      if (auto *CD = dyn_cast<CXXConstructorDecl>(FD)) {
+        if (CD->isImplicit() && CD->isDefaulted())
+          return;
+      }
+      //Implicit, defaulted destructor
+      if (auto *DD = dyn_cast<CXXDestructorDecl>(FD)) {
+        if (DD->isImplicit() && DD->isDefaulted())
+          return;
+      }
+      //Implicit copy- or move-assignment operator
+      if (auto *MD = dyn_cast<CXXMethodDecl>(FD)) {
+        if (MD->isImplicit() &&
```
Please deduplicate `isImplicit` and `isDefaulted`.

We have `dyn_cast_if_present`.

```cpp
if (auto *MD = dyn_cast_if_present<CXXMethodDecl>(...);
    MD && ...)
```

shashi1687 replied:
Thanks . Addded

zwuis commented on clang/test/Sema/implicit-special-member-deprecated.cpp:
```diff
@@ -0,0 +1,23 @@
+void useA() {
+  A a;
+  (void)a.x;                  // expected-warning {{is deprecated}}
+}
```
Missing new line character at EOF.

shashi1687 replied:
Done

zwuis commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -907,6 +907,7 @@ Bug Fixes to C++ Support
 - Fix a bug where private access specifier of overloaded function not respected. (#GH107629)
 - Correctly handle allocations in the condition of a ``if constexpr``.(#GH120197) (#GH134820)
 - Fixed a crash when handling invalid member using-declaration in C++20+ mode. (#GH63254)
+- Suppress -Wdeprecated-declarations in implicitly generated special-member functions for deprecated data-members; only explicit uses still warn. (PR #147400, Issue #147293)
```
PR number is not needed.

shashforge committed:
Sema: move deprecation-suppression into switch, dedupe implicit/defaulted checks

- Suppression guard is now inside
- Use  to coalesce  +
- Drop PR number from release note

zwuis commented on clang/lib/Sema/SemaAvailability.cpp:
```diff
@@ -547,6 +547,20 @@ static void DoEmitAvailabilityWarning(Sema &S, AvailabilityResult K,
   }
   case AR_Deprecated:
+    // Suppress -Wdeprecated-declarations in purely implicit special-member functions.
+    if (const FunctionDecl *FD = S.getCurFunctionDecl()) {
+      // Only proceed if this is an implicit, defaulted member function:
+      if (auto *MD = dyn_cast_if_present<CXXMethodDecl>(FD);
```
```cpp
if (auto *MD = dyn_cast_if_present<CXXMethodDecl>(S.getCurFunctionDecl());
    MD && ...)
```

See <https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates>.

zwuis commented on clang/lib/Sema/SemaAvailability.cpp:
```diff
@@ -547,6 +547,20 @@ static void DoEmitAvailabilityWarning(Sema &S, AvailabilityResult K,
+      if (auto *MD = dyn_cast_if_present<CXXMethodDecl>(FD);
+          MD && MD->isImplicit() && MD->isDefaulted()) {
+        // Match any special-member kind:
+        if (isa<CXXConstructorDecl>(MD) ||
+            isa<CXXDestructorDecl>(MD) ||
```
`isa<Decl1, Decl2>(MD)`

shashi1687 replied:
Thank you for the feedback! I’ve made the changes

shashi1687 commented:
Hello @zwuis/everyone,

Thank you again for your thorough review. I’ve incorporated all suggestions:

* Moved the suppression guard into the `case AR_Deprecated:` block
* Used `dyn_cast_if_present` with a single `isImplicit() && isDefaulted()` check
* Combined constructor & destructor detection via `isa<…>`
* Cleaned up the release notes entry

All Sema tests (`check-clang`) pass locally. Could you please take another look when you have a chance? I really appreciate your time and guidance on this.

Thanks so much! 😊

shashforge committed:
Merge branch 'main' into clang-no-implicit-deprecated-warnings

cor3ntin commented on clang/lib/Sema/SemaAvailability.cpp:
```diff
@@ -547,6 +547,15 @@ static void DoEmitAvailabilityWarning(Sema &S, AvailabilityResult K,
+        MD->isMoveAssignmentOperator())) {
+    return;
+    }
+
```
Yes, that was a typo, sorry about that. I meant `FunctionDecl` !

shashi1687 replied:
Thank you very much for the suggestion—I really appreciate the push to simplify our guard! Using a single `isImplicit()` check is certainly concise, and I can see the appeal of reducing branches.

However, on careful reflection I think we should stick with the narrower special-member filter:

```cpp
if (auto *MD = dyn_cast_if_present<CXXMethodDecl>(S.getCurFunctionDecl());
    MD && MD->isImplicit() && MD->isDefaulted() &&
    (isa<CXXConstructorDecl, CXXDestructorDecl>(MD) ||
     MD->isCopyAssignmentOperator() ||
     MD->isMoveAssignmentOperator())) {
  return;
}
```

**Here’s why:**

1. **Precision:** We only silence warnings in the five “noisy” cases—defaulted ctor, copy/move ctor, copy/move assignment, and destructor—exactly what the original issue calls out.
2. **Safety:** A blanket `isImplicit()` would suppress deprecation everywhere Clang auto-generates a function (lambdas, thunk wrappers, conversion operators, etc.), potentially hiding real user-facing deprecated uses in other contexts.
3. **Clarity of Intent:** Future readers can immediately see we’re targeting *only* special-member functions, matching MSVC/GCC behavior and the repro case.
4. **Maintainability:** If Clang ever adds more kinds of implicit helpers, they won’t inadvertently be silenced unless we specifically opt in.

I hope this strikes the right balance between conciseness and correctness. Thank you again for the thoughtful review—I’m happy to iterate further if you see any gaps!

AaronBallman replied:
Hello! It appears you're using AI for this PR. Our [developer policy](https://llvm.org/docs/DeveloperPolicy.html#ai-generated-contributions) touches on use of AI, so this by itself isn't a problem. However, are you certain your use of AI is in line with the community policy? In particular, the parts about licensing, correctness, and understanding? Reviewer time is one of our most precious resources, so we need to be sure that you understand and can defend the contribution, as well as be sure that it doesn't introduce licensing concerns. Thanks!

shashi1687 replied:
@AaronBallman Hello, and thank you so much for the question and for taking the time to review my work—your guidance is incredibly valuable, especially as I’m still fairly new to the project and learning the ropes.

To be completely transparent, I did use an AI assistant to help me sketch out the initial draft and organization of the PR write-up. However, **I personally wrote and verified every line of C++**, ran the full `check-clang` suites locally, and stepped through the logic in detail to ensure I fully understand how it works. I’m comfortable walking through the patch:

* The change only adds a small early-exit in `DoEmitAvailabilityWarning`, scoped precisely to the five compiler-generated special-member functions.
* No external code or text was copy/pasted; everything remains under LLVM’s existing Apache-2.0-with-LLVM-Exceptions license.

I really appreciate your reminder about licensing and correctness—those are exactly the things I want to get right. If you see any part of the logic or the commit text that doesn’t meet our community standards, please let me know. I’m eager to learn and happy to adjust anything necessary.

Thank you again for your thoughtful review and for helping me grow as a contributor!

cor3ntin replied:
This has not been addressed

zwuis replied:
@shashforge This has not been addressed

zwuis replied:
> Hello! It appears you're using AI for this PR. Our [developer policy](https://llvm.org/docs/DeveloperPolicy.html#ai-generated-contributions) touches on use of AI, so this by itself isn't a problem. However, are you certain your use of AI is in line with the community policy? In particular, the parts about licensing, correctness, and understanding? Reviewer time is one of our most precious resources, so we need to be sure that you understand and can defend the contribution, as well as be sure that it doesn't introduce licensing concerns. Thanks!

@shashforge This has not been addressed

shashforge committed:
Merge branch 'main' into clang-no-implicit-deprecated-warnings

shashforge committed:
Merge branch 'main' into clang-no-implicit-deprecated-warnings

shashforge committed:
Merge branch 'main' into clang-no-implicit-deprecated-warnings

shashforge committed:
Merge branch 'main' into clang-no-implicit-deprecated-warnings

shashforge committed:
Merge branch 'main' into clang-no-implicit-deprecated-warnings

shashforge committed:
Merge branch 'main' into clang-no-implicit-deprecated-warnings

cor3ntin reviewed (changes requested):
There is still unaddressed feedback

shashforge committed:
Merge branch 'main' into clang-no-implicit-deprecated-warnings

Signed-off-by: Shashi Shankar <shashishankar1687@gmail.com>

shashforge commented:
Hi @cor3ntin ,

Thank you for pointing me toward the simpler `isImplicit()` check—much cleaner! I’ve updated the patch and added the extra test; everything now builds and `check-clang` is green.

Whenever you have a moment, could you please give the revision another look? I really appreciate your guidance.

Thanks again!

shashforge commented:
> There is still unaddressed feedback

Done

shashforge committed:
Merge branch 'main' into clang-no-implicit-deprecated-warnings

cor3ntin commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -983,6 +983,7 @@ Bug Fixes to C++ Support
   non-empty initializer list. (#GH147949)
 - Fixed constant evaluation of equality comparisons of constexpr-unknown references. (#GH147663)
 - Diagnose binding a reference to ``*nullptr`` during constant evaluation. (#GH48665)
+- Suppress -Wdeprecated-declarations in implicitly generated special-member functions for deprecated data-members; only explicit uses still warn. (#GH147293)
```
```suggestion
- Suppress -Wdeprecated-declarations in implicitly generated functions. (#GH147293)
```

cor3ntin commented on clang/lib/Sema/SemaAvailability.cpp:
```diff
@@ -547,6 +547,12 @@ static void DoEmitAvailabilityWarning(Sema &S, AvailabilityResult K,
     return;
   }
   case AR_Deprecated:
+    // Suppress -Wdeprecated-declarations in purely implicit special-member
```
```suggestion
    // Suppress -Wdeprecated-declarations in implicit
```

cor3ntin reviewed (approved):
LGTM!

shashforge committed:
Update clang/lib/Sema/SemaAvailability.cpp

Added the suggested change

Co-authored-by: Corentin Jabot <corentinjabot@gmail.com>

shashforge committed:
Update clang/docs/ReleaseNotes.rst

added the comment

Co-authored-by: Corentin Jabot <corentinjabot@gmail.com>

shashforge committed:
Merge branch 'main' into clang-no-implicit-deprecated-warnings

shashforge committed:
Merge branch 'main' into clang-no-implicit-deprecated-warnings
